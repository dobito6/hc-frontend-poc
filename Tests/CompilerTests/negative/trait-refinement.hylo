//! no-std

trait Base {
  fun f()
}

trait Derived {
  static given let _: Base<Self>
  fun g() { self.f() }
}

fun ghi<T where T is Derived>(_ x: T) {
  x.g()
  x.f()
}

struct A<T> {}
struct B<T> {}

given <T> A<T> is Base {
  fun f() {}
}

// This one is OK.
given <T> A<T> is Derived {}

// This one is missing an instance of `Base<B<T>>`.
given <T> B<T> is Derived {}
