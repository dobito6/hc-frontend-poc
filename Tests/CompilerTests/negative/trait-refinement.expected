negative/trait-refinement.hylo:28.1-6: error: no given instance of 'Base<B<T>>' in this scope
given <T> B<T> is Derived {}
~~~~~
